### PR TITLE
Feature/match monitor

### PR DESCRIPTION
**Pull request by maciejburda**
Added simple polling match monitor

### CODE REVIEW

File: sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt [hunk 1]
```diff
@@ -0,0 +1,51 @@
+package io.matchmore.sdk.monitoring
+
+import io.matchmore.sdk.AlpsManager
+import io.matchmore.sdk.api.async
+import io.matchmore.sdk.api.models.Device
+import io.matchmore.sdk.api.models.Match
+import java.util.*
+
+interface MatchMonitorDelegate {
```
Comment (kubatatami) [inline sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt]: MatchMonitorListener will be more "Android"

File: sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    fun didReceiveMatches(matches: Set<Match>, forDevice: Device)
```
Comment (kubatatami) [inline sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt]: onReceiveMatches will be more "Android"

File: sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+class MatchMonitor(private val manager: AlpsManager, private val delegate: MatchMonitorDelegate) {
```
Comment (kubatatami) [inline sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt]: Can you add example of usage in tests or example app ?

File: sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        }
+    }
+
+    private fun getMatchesForDevice(device: Device) {
+        device.id?.let {
+            manager.apiClient.matchesApi.getMatches(it).async({
+                val union = deliveredMatches.union(it.toSet())
+                if (!(deliveredMatches == union)) {
+                    deliveredMatches = union.toMutableSet()
+                    delegate?.didReceiveMatches(union, device)
```
Comment (kubatatami) [inline sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt]: delegate - You check optionality but field declaration in constructor is not optional.

Comment (kubatatami) [inline sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt]: I didn't test it but probably you can write it:
`
val newMatches = it - deliveredMatches
if (newMatches.isNotEmpty()) {
    deliveredMatches.addAll(newMatches)
    delegate.didReceiveMatches(newMatches.toSet(), device)
}
`
The same lines but clearer? What do you think?

File: sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import io.matchmore.sdk.api.models.Match
+import java.util.*
+
+interface MatchMonitorDelegate {
+    fun didReceiveMatches(matches: Set<Match>, forDevice: Device)
+}
+
+class MatchMonitor(private val manager: AlpsManager, private val delegate: MatchMonitorDelegate) {
+    var monitoredDevices = mutableSetOf<Device>()
+    var deliveredMatches = mutableSetOf<Match>()
```
Comment (kubatatami) [inline sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt]: Can be private ?

File: sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+    private var timer: Timer? = null
+
+    fun startPollingMatchesA() {
+        val period: Long = 5000 // repeat every 10 sec.
```
Comment (kubatatami) [inline sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt]: 5 or 10 ?:) Better to move it to companion object.

File: sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+            }
+        }, 0, period)
+    }
+
+    fun stopPollingMatches() {
+        timer?.cancel()
+        timer = null
+    }
+
+    private fun getMatches() {
```
Comment (kubatatami) [inline sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt]: You can write it in one line if you want:
`private fun getMatches() = monitoredDevices.forEach { getMatchesForDevice(it) }`

File: sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt (hunk 5 above)
Comment (maciejburda) [reply sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt]: I want them both to be public. Devices will be set, and delivered matches are for a case where you want your all matches (not only recent ones)

File: sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt (hunk 6 above)
Comment (maciejburda) [reply sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt]: Good catch :P

File: sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt (hunk 4 above)
Comment (maciejburda) [reply sdk/src/main/java/io/matchmore/sdk/monitoring/MatchMonitor.kt]: I was thinking to actually make it multiple delegates/observers. What is best way to do it?

**Comment by maciejburda** [comment]
I will merge it now and add tests later on @kubatatami

**Pull request merged**